// @servisbot/servisbot-api-sdk/src/Maker/API.js
/**
 * API - A Maker API request - can be used with resolvers or Maker.BusinessIntent fulfillment
 * @class
 *
 * @param {Object} api
 * @param {string} api.endpoint - The URl of the resource we are requesting. Use ${param} in combination with Maker.APIParam.URLParam to inject params.
 * @param {Object} api.requestMapping - A map of Maker.APIParam. Maps values into the request from context before being sent.
 * @param {Object} api.responseMapping - A map of Maker.APIParam. Maps values from the request into context after a response has been received.
 * @param {Object} api.headers - A map of headers. Use ${param} in combination with Maker.APIParam.HeaderParam to inject content.
 * @param {string} api.method - HTTP method of the request
 * @param {Object} api.body - HTTP request body for the request. Hardcode content here, or inject content later by setting params on api.requestMapping
 * @returns {Maker.API}
 */
class API {
  constructor({
    alias, endpoint, requestMapping, responseMapping, headers, body, method, errors
  }) {
    this.alias = alias;
    this.endpoint = endpoint;
    this.requestMapping = requestMapping || {};
    this.responseMapping = responseMapping || {};
    this.headers = headers;
    this.method = method || 'get';
    this.body = body || {};
    this.errors = errors || [];
    return this;
  }
}

module.exports = API;
