// @servisbot/servisbot-api-sdk/src/Maker/Orchestrator.js
const validation = require('./util/lookupSlots');
const softBracesCheck = require('./util/softBracesCheck');
/**
 * Orchestrator - maps UserIntent through to BusinessIntent
 * @class
 *
 * @param {Object} orchestrator
 * @param {Maker.BusinessIntent} orchestrator.businessIntent - Business Intent being created
 * @param {Maker.ResponseVoice} orchestrator.responseVoice - Response voice for this bot to use
 * @param {Maker.PromptStrategy} orchestrator.promptStrategy - Prompt strategy in use
 * @param {Object} orchestrator.contextMapping - Context to apply to every request
 * @param {Maker.UserIntent[]} orchestrator.userIntents - userIntents to map
 *
 * @returns {Maker.Orchestrator} a new orchestrator instance
 */
class Orchestrator {
  constructor({
    businessIntent, responseVoice, promptStrategy, contextMapping, userIntents, language
  }) {
    // Validate that UserIntents are covered within the BusinessIntent
    validation.userIntentSlotsCanBeFufilled(userIntents, businessIntent);
    this.businessIntent = businessIntent;
    this.responseVoice = responseVoice;
    this.promptStrategy = promptStrategy;
    this.contextMapping = contextMapping;
    this.userIntents = userIntents;
    this.language = language;
    softBracesCheck(this);
  }
}

module.exports = Orchestrator;
