// @servisbot/servisbot-api-sdk/src/Maker/Resolver.js
/**
 * Resolver - specifies means to fill a Business Slot
 * @class
 *
 * @param {Object} resolver
 * @param {String} resolver.name - Title for this resolver
 * @param {Maker.SlotType} resolver.slotType - The type of slot this resolver fills
 * @param {Boolean} resolver.defaultResolver - Uses this resolver when no slot was found in the utterance, instead of eliciting a value. Will use either a defaultValue or apiDefinition to fulfill.
 * @param {*} resolver.defaultValue - A hard-coded default value to use for this slot when no value was found in the utterance
 * @param {Maker.API} resolver.apiDefinition - The API called on filling this slot - can be used to fill more information in context, to validate, or to retrieve default values.
 *
 * @returns {Maker.Resolver}
 */
class Resolver {
  constructor({
    name, slotType, apiDefinition, defaultResolver, defaultValue, disambiguates, defaultApi
  }) {
    this.name = name;
    this.defaultResolver = (defaultResolver === true);
    this.defaultValue = defaultValue;
    this.slotType = slotType;
    this.apiDefinition = apiDefinition;
    this.defaultApi = defaultApi;
    this.disambiguates = disambiguates;
    return this;
  }
}

module.exports = Resolver;
