// @servisbot/servisbot-api-sdk/src/ServisBotApiSdk.js
const BotArmyRuntimeLib = require('./BotArmyRuntime');
const BotArmyLib = require('./BotArmy');
const MakerLib = require('./Maker/');
const BaaSLib = require('./BaaS/');
const OperatorLib = require('./Operator/');
const SherpaGatewayLib = require('./SherpaGateway');
/**
 * @typedef {Object} BotArmyRuntime
 * @property {createConversation} createConversation
 */
const ServisBotApiSdk = (fetch, Promise, Logger, shortid) => ({ // eslint-disable-line
  /**
   * Initialises the SDK with the given URL and API key
   *
   * @param {string} url
   * @param {string} key
   */
  init(url, key) {
    if (!url || !key) return Promise.reject(new Error('Url and/or key not valid'));
    this.BotArmyRuntime.init(url, key);
    this.BotArmy.init(url, key);
    this.SherpaGateway.init(url, key);
    return Promise.resolve();
  },
  BotArmyRuntime: BotArmyRuntimeLib(fetch).BotArmyRuntime,
  BotArmy: BotArmyLib(fetch).BotArmy,
  Maker: MakerLib().Maker,
  BaaS: BaaSLib().BaaS,
  Operator: OperatorLib(Logger).Operator,
  SherpaGateway: SherpaGatewayLib(fetch).SherpaGateway,
});

module.exports = ServisBotApiSdk;
